src/v/model/record_batch_reader.cc

```diff
@@ -110,12 +110,8 @@ record_batch_reader make_memory_record_batch_reader(storage_t batches) {
 
 record_batch_reader
 make_foreign_memory_record_batch_reader(record_batch_reader::data_t data) {
-    auto batches = std::make_unique<record_batch_reader::data_t>(
-      std::move(data));
-    return make_memory_record_batch_reader(record_batch_reader::foreign_data_t{
-      .buffer = ss::make_foreign(std::move(batches)),
-      .index = 0,
-    });
+    return make_foreign_record_batch_reader(
+      make_memory_record_batch_reader(std::move(data)));
```

Comment:
something about this doesn't quite seem right. this looks like the foreign record batch reader will jump back over to the source core to load the slice which would be unnecessary since this is heap allocated data.

iirc you are trying to get the `impl` itself to be deallocated on the correct core? if so, would we instead want to teach `class record_batch_reader` to store a `std::unique_ptr<impl> _impl;` or the equivalent foreign wrapper?